openmls/src/group/core_group/apply_proposals.rs

```diff
@@ -95,18 +96,19 @@ impl CoreGroup {
                     // This should not happen with validated proposals
                     _ => return Err(LibraryError::custom("Update proposal from non-member").into()),
                 };
-                let leaf_node: OpenMlsLeafNode = if sender_index == self.tree.own_leaf_index() {
-                    let own_leaf_node = match leaf_nodes
-                        .iter()
-                        .find(|&leaf_node| leaf_node.leaf_node() == update_proposal.leaf_node())
-                    {
-                        Some(leaf_node) => leaf_node,
-                        // We lost the LeafNode apparently
-                        None => return Err(ApplyProposalsError::MissingLeafNode),
-                    };
-                    own_leaf_node.clone()
-                } else {
-                    update_proposal.leaf_node().clone().into()
+                let leaf_node: OpenMlsLeafNode = match own_leaf_index {
+                    Some(leaf_index) if leaf_index == sender_index => {
+                        let own_leaf_node = match leaf_nodes
+                            .iter()
+                            .find(|&leaf_node| leaf_node.leaf_node() == update_proposal.leaf_node())
+                        {
+                            Some(leaf_node) => leaf_node,
+                            // We lost the LeafNode apparently
+                            None => return Err(ApplyProposalsError::MissingLeafNode),
+                        };
+                        own_leaf_node.clone()
+                    }
+                    _ => update_proposal.leaf_node().clone().into(),
                 };
                 diff.update_leaf(leaf_node, sender_index);
```

Comment:
`update_leaf` should probably get a `LeafNode` instead, then we safe us some unnecessary conversions.

Reply:
I'm not sure there is something we can save here until we have ripped out the `OpenMlsLeafNode` abstraction. That's probably possible (it only contains an optional `LeafNodeIndex` that we can probably get in some other way), but not quite yet.

Reply:
But `update_proposal` isn't using the index from the `leaf_node`. So the `into` in `update_proposal` is unnecessary, which would allow you to drop the `into` in line 113.
But not too important for now.

Reply:
I'm confused. `update_leaf` puts the new leaf into the tree, which means that at some point whatever input it takes needs to become a `TreeSyncLeafNode` via an `OpenMlsLeafNode`. `update_proposal` contains a `LeafNode`. So those `into` calls have to happen at some point. If I remove them here, I have to add them inside of `update_leaf`. Is that what you mean?